mitodl/mitxonline
# 904: tracking audit enrollments for mitx online courses in hubspot

#### Pre-Flight checklist

#### What are the relevant tickets?
https://github.com/mitodl/hq/issues/904

#### What's this PR do?
1. When a user enrolls into a course, a Deal is created in HubSpot along with Line items that contain information on the `enrollment_mode` and `change_status` related to the user's enrollment in the associated course run.
2. When a user proceeds to upgrade or pay for the certificate track (verified) for a course run, the existing `PendingOrder` is used for initializing the checkout workflow.
3. At the time of checkout, if there are no PendingOrders associated with the user that contain Line items matching the products found in the cart, then a new PendingOrder is created.
4. When a user unenrolls from a course run, the Line item in HubSpot is updated.
5. When a user has a FulfilledOrder but their associated CourseRunEnrollment is not synced with Edx, and the user attempts to enroll into the course again (From the course about page) then no new PendingOrder shall be created.

#### How should this be manually tested?

1. Have a Course, CourseRun, User, Product, Discount, link the Discount with the Product.
2. Enroll your user into the Course from the course about page.
3. Verify that a PendingOrder is created after selecting an option on the upsell modal.
4. Verify that a HubSpot Deal is created and the associated Line displays the correct enrollment information for enrollment mode and change status.
5. Ensure that Edx is not running or is not connected with your MITx Online instance.
6. Upgrade and pay for your course enrollment using a Discount code.
7. Verify that a FulFilledOrder exists.
8. Verify that the HubSpot Deal from step 4 has been updated to reflect a closed deal.
9. Verify that the HubSpot Deal's Line from step 4 has been updated to reflect the `verified` enrollment mode.
10. From the course about page, attempt to enroll in the course again.
11. Verify that no PendingOrder is created from step 10.
12. Unenroll from your course.
13. Ensure that the HubSpot Deal's Line from step 4 has been updated to reflect the change_status of the CourseRunEnrollment.

#### Where should the reviewer start?

1. Have the required access to HubSpot for testing MITx Online locally.  Ask if you need it.
14. Configure HubSpot with your local instance of MITx Online.

#### Any background context you want to provide?

- There are a few places where I've removed unused imports that were not related to this work.
- I refactored `create_from_basket()` and `create_from_product()` to both call a private method since much of their functionality was shared.
